amethyst_renderer/src/pipe/pass.rs

```diff
@@ -4,66 +4,76 @@
 use std::fmt::{Debug, Formatter, Result as FmtResult};
 use std::marker::PhantomData;
 
-use rayon::iter::{IntoParallelIterator, ParallelIterator};
-use rayon::iter::internal::UnindexedConsumer;
+use rayon::iter::IntoParallelIterator;
 use rayon_core;
+use specs::SystemData;
 
 use error::Result;
 use pipe::{Effect, NewEffect, Target};
 use types::{Encoder, Factory};
 
-
-#[derive(Clone)]
-struct Dispenser<'a> {
-    encoders: *mut [Encoder],
+pub struct Dispenser<'a, T> {
+    slice: *mut [T],
     pd: PhantomData<&'a ()>,
 }
 
-impl<'a> Dispenser<'a> {
-    fn new(encoders: &'a mut [Encoder]) -> Self {
+impl<'a, T> Clone for Dispenser<'a, T> {
+    fn clone(&self) -> Self {
         Dispenser {
-            encoders: encoders,
+            slice: self.slice,
             pd: PhantomData,
         }
     }
+}
 
-    fn get(&mut self) -> &mut Encoder {
-        let encoders = unsafe { &mut *self.encoders };
-        let count = encoders.len();
-        encoders
-            .get_mut(self.index())
-            .expect(&format!("Not enough encoders. Index: {}, Encoders count: {}", self.index(), count))
+impl<'a, T> Dispenser<'a, T> {
+    fn new(slice: &'a mut [T]) -> Self {
+        Dispenser {
+            slice: slice,
+            pd: PhantomData,
+        }
     }
 
-    fn index(&mut self) -> usize {
+    fn index(&self) -> usize {
         rayon_core::current_thread_index().expect("Should be called from ThreadPool")
     }
+
+    // FIXME: make it accept `&mut self` when upgrading to rayon v0.8
+    pub unsafe fn get(&self) -> &mut T {
+        let slice = &mut *self.slice;
+        let count = slice.len();
+        slice
+            .get_mut(self.index())
+            .expect(&format!("Not enough objects. Index: {}, Dispenser count: {}", self.index(), count))
+    }
 }
 
+unsafe impl<'a, T: 'a> Send for Dispenser<'a, T> where &'a mut T: Send {}
+unsafe impl<'a, T: 'a> Sync for Dispenser<'a, T> where &'a mut T: Send {}
 
-unsafe impl<'a> Send for Dispenser<'a> {}
-unsafe impl<'a> Sync for Dispenser<'a> {}
 
+pub type Encoders<'a> = Dispenser<'a, Encoder>;
+pub type Effects<'a> = Dispenser<'a, Effect>;
 
-pub struct Only<T>(PhantomData<T>);
-pub struct All<T>(PhantomData<T>);
-pub struct Each<T>(PhantomData<T>);
+pub trait PassCompiler {
+    fn compile(&self, effect: NewEffect) -> Result<Effect>;
+}
 
-pub trait Hint {
-    type Type;
+pub trait PassApply<'a> {
+    type Apply: IntoParallelIterator<Item=()>;
 }
 
-pub trait PassCompiler {
-    fn compile(&self, effect: NewEffect) -> Result<Effect>;
+pub trait PassData<'a> {
+    type Data: SystemData<'a> + Send;
```

Comment:
Are these "parent" traits to `SystemData` for extensibility? Or is it just restricting the types more?